nexus/src/app/background/tasks/saga_recovery.rs

```diff
@@ -0,0 +1,916 @@
+        // unplug the datastore partway through so that the later log entries
+        // don't get written.  Note that the unplugged datastore completes
+        // operations successfully so that the saga will appeaer to complete
+        // successfully.
+        uctx.do_unplug.store(true, Ordering::SeqCst);
+        let (_, result) = run_test_saga(&uctx, &sec_client).await;
+        let output = result.kind.unwrap();
+        assert_eq!(output.lookup_node_output::<i32>("n1_out").unwrap(), 1);
+        assert_eq!(output.lookup_node_output::<i32>("n2_out").unwrap(), 2);
+        assert_eq!(uctx.n1_count.load(Ordering::SeqCst), 1);
+        assert_eq!(uctx.n2_count.load(Ordering::SeqCst), 1);
+
+        // Simulate a crash by terminating the SEC and creating a new one using
+        // the same storage system.
+        //
+        // Update uctx to prevent the storage system from detaching again.
+        sec_client.shutdown().await;
+        let sec_client = steno::sec(sec_log, storage.clone());
+        uctx.storage.set_unplug(false);
+        uctx.do_unplug.store(false, Ordering::SeqCst);
+
+        // Use our background task to recover the saga.  Observe that it re-runs
+        // operations and completes.
+        let sec_client = Arc::new(sec_client);
+        let (_, sagas_started_rx) = tokio::sync::mpsc::unbounded_channel();
+        let mut task = SagaRecovery::new(
+            db_datastore.clone(),
+            sec_id,
+            saga_recovery_opctx,
+            uctx.clone(),
+            sec_client.clone(),
+            registry_create(),
+            sagas_started_rx,
+        );
+
+        let Some((completion_future, last_pass_success)) =
+            task.activate_internal(&opctx).await
+        else {
+            panic!("saga recovery failed");
+        };
+
+        assert_eq!(last_pass_success.nrecovered, 1);
+        assert_eq!(last_pass_success.nfailed, 0);
+        assert_eq!(last_pass_success.nskipped, 0);
+
+        // Wait for the recovered saga to complete and make sure it re-ran the
+        // operations that we expected it to.
+        completion_future
+            .await
+            .expect("recovered saga to complete successfully");
+        assert_eq!(uctx.n1_count.load(Ordering::SeqCst), 2);
+        assert_eq!(uctx.n2_count.load(Ordering::SeqCst), 2);
+
+        // Test cleanup
+        drop(task);
+        let sec_client = Arc::try_unwrap(sec_client).unwrap();
+        sec_client.shutdown().await;
+        db.cleanup().await.unwrap();
+        logctx.cleanup_successful();
+    }
+
+    // Tests that a saga that has finished (as reflected in the database state)
+    // does not get recovered.
+    #[tokio::test]
+    async fn test_successful_saga_does_not_replay_during_recovery() {
+        // Test setup
+        let logctx = dev::test_setup_log(
+            "test_successful_saga_does_not_replay_during_recovery",
+        );
+        let log = logctx.log.new(o!());
```

Comment:
GitHub's syntax highlighting seems to have really choked on something here, but I can't see any obvious syntax errors...seems weird: 
![image](https://github.com/user-attachments/assets/b155de79-ea25-4dda-ad54-ab521424e463)

probably not a huge deal though

Reply:
Yeah, I'm not sure what's going on there but Rust, rustfmt, and vim have no issues with this so I'm not sure it's worth debugging GitHub.